scripts/apps/monitoring/MultiActionBarReact.tsx

```diff
@@ -0,0 +1,141 @@
+import {sortByDisplayPriority} from 'core/helpers/sortByDisplayPriority';
+
+export interface IArticleActionBulkExtended extends IArticleActionBulk {
+    // this is possible for all extensions since they don't depend on external state
+    // most of core actions are relying on service state which would be destroyed if we closed the bar too early
+    // when `canAutocloseMultiActionBar` set to false, core actions close it themselves when they are done
+    canAutocloseMultiActionBar: boolean;
+}
+
+interface IProps {
+    context: 'archive' | 'ingest';
+    articles: Array<IArticle>;
+    compact: boolean;
+    getCoreActions(articles: Array<IArticle>): Array<IArticleActionBulkExtended>;
+    hideMultiActionBar(): void;
+}
+
+interface IState {
+    actions?: Array<IArticleActionBulkExtended>;
+}
+
+function getActionsBulkFromExtensions(articles): Promise<Array<IArticleActionBulkExtended>> {
+    const getActionsBulk
+    : Array<IExtensionActivationResult['contributions']['entities']['article']['getActionsBulk']>
+    = flatMap(
+        Object.values(extensions).map(({activationResult}) => activationResult),
+        (activationResult) =>
+            activationResult.contributions != null
+            && activationResult.contributions.entities != null
+            && activationResult.contributions.entities.article != null
+            && activationResult.contributions.entities.article.getActionsBulk != null
+                ? activationResult.contributions.entities.article.getActionsBulk
+                : [],
+    );
+
+    return Promise.all(
+        getActionsBulk.map((getPromise) => getPromise('include', articles)),
+    ).then((res) => flatMap(res).map((action) => ({...action, canAutocloseMultiActionBar: true})));
+}
+
+export class MultiActionBarReact extends React.Component<IProps, IState> {
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {};
+    }
+    componentDidMount() {
+        getActionsBulkFromExtensions(this.props.articles).then((actionsBulkFromExtensions) => {
+            this.setState({
+                actions: sortByDisplayPriority(
+                    this.props.getCoreActions(this.props.articles).concat(actionsBulkFromExtensions),
+                ),
+            });
+        });
+    }
+    componentDidUpdate(prevProps) {
+        // update when more items are selected / deselected
+        if (prevProps !== this.props) {
+            getActionsBulkFromExtensions(this.props.articles).then((actionsBulkFromExtensions) => {
+                this.setState({
+                    actions: sortByDisplayPriority(
+                        this.props.getCoreActions(this.props.articles).concat(actionsBulkFromExtensions),
+                    ),
+                });
+            });
+        }
+    }
+    render() {
+        if (this.state.actions == null) {
+            return null;
+        }
+
+        const onTrigger = (action: IArticleActionBulkExtended) => {
```

Comment:
Shouldn't this be a method instead?

Reply:
I don't like having methods if I can avoid it. The issue with methods is that they can be called by any other method in the class and any code outside it that has a reference to the class. That's a lot of stuff to check in order to understand which code the method can be called by. We have TypeScript now and can use "find usages" command which makes it easier, but I still like the simplicity of having a function instead of a class method and being able to immediately understand that no other code can call it.

There is one issue though. Since I defined the function in the render method, it will get recreated every time render runs. Which means child components I pass the function to, won't be able to understand, at least by comparing the reference, that it's the same function, and will trigger a re-render. To avoid this, it's better to define the function outside of the component which I'll do.

Reply:
I understand. We could also start to use `private` methods to avoid instance calls